File: Python/0406-queue-reconstruction-by-height.py
# Time:  O(n^2)
# Space: O(n)

class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        people.sort(key=lambda h_k: (-h_k[0], h_k[1]))
        result = []
        for p in people:
            result.insert(p[1], p)
        return result

File: C++/0406-queue-reconstruction-by-height.cpp
// Time:  O(n^2)
// Space: O(n)

class Solution {
public:
    vector<vector<int>> reconstructQueue(vector<vector<int>>& people) {
        sort(people.begin(), people.end(), [](const vector<int> &a, const vector<int> &b)
             {return a[0] == b[0] ? a[1] < b[1] : a[0] > b[0];});

        vector<vector<int>> result;
        for (const auto &p: people) {
            // insert at index k
            result.insert(result.begin() + p[1], p);
            // or result.emplace()
        }
        return result;
    }
};
